web_poet/mixins.py

```diff
@@ -78,3 +81,30 @@ def urljoin(self, url: str) -> str:
         url and baseurl of the response"""
         # FIXME: move it to HttpResponse
         return urljoin(self.base_url, url)
+
+
+class _UrlMixin(abc.ABC):
+    """Common shortcut methods for working with the URL of HTTP requests and
+    responses.
+
+    This mixin could be used with Page Object base classes.
+
+    It requires the ``url`` attribute to be present.
+    """
+
+    _cached_base_url = None
+    url: Union[str, RequestUrl, ResponseUrl]
+
+    @property
+    def _base_url(self) -> str:
+        if self._cached_base_url is None:
+            text = getattr(self, "text", "")[:4096]
```

Comment:
It seems having "text" attribute is another requirement for this mixin - should we document it? 

In SelectableMixin it's solved with a private abc method, this allows the name of the attribute to change between classes (e.g. ResponseShortcutsMixin uses "html"). But this may be not important here.

Reply:
I will document it.

SelectableMixin’s solution makes sense for a public mixin, but I don‘t think we need to do that here, this being a private mixin. We can always do it later on if we want our built-in classes to expose `text` with different names.